File: src/doux_planning/api/sandbox.py
from __future__ import annotations

from typing import Any

from sqlalchemy.orm.attributes import flag_modified

from doux_planning.engine import EngineResult, Shift, _attempt_key
from doux_planning.hydrate import hydrate_delivered_cycle
from doux_planning.planning import (
    EmptyHistoryError,
    FillSlot,
    PlanningStore,
    PreviewImpact,
    PreviewProposal,
    SandboxSnapshot,
)
from doux_planning.types import Team, WarningSeverity
from doux_planning.warnings import Warning

RESTAURANT_ID = "saint-cloud"
GESTURES = {"retune", "replace", "swap", "fill"}
SCORE_FIELDS = ("empty", "interdit", "hours_miss", "souhait", "below_role", "overqualification")

_store: PlanningStore | None = None
_recaps: list[dict[str, Any]] = []
_restored = False


def reset_runtime(*, clear_db: bool = True) -> None:
    global _store, _recaps, _restored
    _store = None
    _recaps = []
    _restored = False
    if clear_db:
        _delete_persisted()


def get_store() -> PlanningStore:
    global _store, _recaps, _restored
    if _store is None:
        _store = PlanningStore()
        _recaps = []
        if not _restored:
            _restored = True
            _restore()
    return _store


def enter_sandbox_state() -> dict[str, Any]:
    store = get_store()
    try:
        store.get(RESTAURANT_ID)
    except KeyError:
        hydrate_delivered_cycle(store, RESTAURANT_ID)
        _recaps.clear()
    else:
        store.enter_sandbox(RESTAURANT_ID, "cycle")
    _persist()
    return sandbox_state()


def current_sandbox_state() -> dict[str, Any]:
    sandbox = _sandbox_or_none()
    if sandbox is None:
        raise LookupError("no sandbox")
    return sandbox_state()


def preview(body: dict[str, Any]) -> dict[str, Any]:
    gesture = body.get("gesture")
    if gesture not in GESTURES:
        raise ValueError("unknown gesture")
    proposals = _preview(get_store(), gesture, body)
    return {"proposals": [_proposal_json(item) for item in proposals]}


def commit(body: dict[str, Any]) -> dict[str, Any]:
    gesture = body.get("gesture")
    if gesture not in GESTURES:
        raise ValueError("unknown gesture")
    store = get_store()
    proposals = _preview(store, gesture, body)
    chosen = _match_proposal(gesture, body, proposals)
    if chosen is None:
        raise LookupError("no proposal")
    store.apply_proposal(RESTAURANT_ID, chosen)
    _recaps.append(_recap_json(gesture, body, chosen, len(_recaps) + 1))
    _persist()
    return sandbox_state()


def undo() -> dict[str, Any]:
    store = get_store()
    store.undo_sandbox(RESTAURANT_ID)
    if _recaps:
        _recaps.pop()
    _persist()
    return sandbox_state()


def discard_sandbox_state() -> dict[str, Any]:
    store = get_store()
    if _sandbox_or_none() is None:
        raise LookupError("no sandbox")
    store.discard_sandbox(RESTAURANT_ID)
    _recaps.clear()
    store.enter_sandbox(RESTAURANT_ID, "cycle")
    _delete_persisted()
    _persist()
    return sandbox_state()


def sandbox_state() -> dict[str, Any]:
    state = get_store().get(RESTAURANT_ID)
    sandbox = state.sandbox
    assert sandbox is not None
    result = sandbox.last_result
    assignments = sandbox.draft.assignments
    warnings = result.warnings if result is not None else ()
    return {
        "sandbox": {"target": "cycle", "history_length": len(sandbox.history)},
        "restaurant": {
            "id": state.identity.id,
            "name": "Saint-Cloud",
            "employees": [_employee_json(person) for person in state.employees],
        },
        "planning": {
            "assignments": [_shift_json(item) for item in assignments],
            "warnings": [_warning_json(item) for item in warnings],
        },
        "score": _score_json(sandbox.draft, result),
        "history": list(_recaps),
    }


def parse_shift(raw: dict[str, Any]) -> Shift:
    required = (
        "employee_id",
        "day_index",
        "weekday",
        "service_id",
        "team",
        "start_minutes",
        "end_minutes",
        "post_level",
    )
    missing = [key for key in required if key not in raw]
    if missing:
        raise ValueError("missing shift fields")
    return Shift(
        employee_id=raw["employee_id"],
        day_index=int(raw["day_index"]),
        weekday=raw["weekday"],
        service_id=raw["service_id"],
        team=Team(raw["team"]),
        start_minutes=int(raw["start_minutes"]),
        end_minutes=int(raw["end_minutes"]),
        post_level=int(raw["post_level"]),
    )


def parse_slot(raw: dict[str, Any]) -> FillSlot:
    required = ("employee_id", "day_index", "weekday", "service_id", "team")
    missing = [key for key in required if key not in raw]
    if missing:
        raise ValueError("missing gesture fields")
    return FillSlot(
        employee_id=raw["employee_id"],
        day_index=int(raw["day_index"]),
        weekday=raw["weekday"],
        service_id=raw["service_id"],
        team=Team(raw["team"]),
    )


def _fill_times(body: dict[str, Any]) -> tuple[int | None, int | None]:
    start = body.get("start_minutes", None)
    end = body.get("end_minutes", None)
    start_present = "start_minutes" in body
    end_present = "end_minutes" in body
    if not start_present and not end_present:
        return None, None
    if start is None and end is None:
        return None, None
    if start is None or end is None:
        raise ValueError("missing gesture fields")
    return int(start), int(end)


def _preview(store: PlanningStore, gesture: str, body: dict[str, Any]) -> list[PreviewProposal]:
    if gesture == "fill":
        slot_raw = body.get("slot")
        if not isinstance(slot_raw, dict):
            raise ValueError("missing gesture fields")
        start, end = _fill_times(body)
        return store.preview_fill(RESTAURANT_ID, parse_slot(slot_raw), start, end)
    shift = parse_shift(body.get("shift") or {})
    if gesture == "retune":
        if "start_minutes" not in body or "end_minutes" not in body:
            raise ValueError("missing gesture fields")
        return store.preview_retune(
            RESTAURANT_ID,
            shift,
            int(body["start_minutes"]),
            int(body["end_minutes"]),
        )
    if gesture == "replace":
        return store.preview_replace(RESTAURANT_ID, shift)
    if gesture == "swap":
        return store.preview_swap(RESTAURANT_ID, shift)
    raise ValueError("unknown gesture")


def _match_proposal(gesture: str, body: dict[str, Any], proposals: list[PreviewProposal]) -> PreviewProposal | None:
    if gesture == "retune":
        return proposals[0] if proposals else None
    if gesture in {"replace", "fill"}:
        if "employee_id" not in body:
            raise ValueError("missing gesture fields")
        employee_id = body["employee_id"]
        for item in proposals:
            if item.employee_id == employee_id:
                return item
        return None
    partner_raw = body.get("partner")
    if not isinstance(partner_raw, dict):
        raise ValueError("missing gesture fields")
    partner = parse_shift(partner_raw)
    for item in proposals:
        if item.partner == partner:
            return item
    return None


def _sandbox_or_none():
    store = get_store()
    try:
        return store.get(RESTAURANT_ID).sandbox
    except KeyError:
        return None


def _employee_json(person) -> dict[str, Any]:
    return {
        "id": person.id,
        "name": person.name,
        "role": {"name": person.role.name, "level": person.role.level, "team": person.role.team.value},
        "team": person.team.value,
    }


def _shift_json(shift: Shift) -> dict[str, Any]:
    return {
        "employee_id": shift.employee_id,
        "day_index": shift.day_index,
        "weekday": shift.weekday,
        "service_id": shift.service_id,
        "team": shift.team.value,
        "start_minutes": shift.start_minutes,
        "end_minutes": shift.end_minutes,
        "post_level": shift.post_level,
        "duration_hours": shift.duration_hours,
    }


def _warning_json(warning: Warning) -> dict[str, Any]:
    return {
        "severity": warning.severity.value,
        "code": warning.code,
        "message": warning.message,
        "employee_id": warning.employee_id,
        "day_index": warning.day_index,
    }


def _score_json(draft, result: EngineResult | None) -> dict[str, Any]:
    if result is None:
        raise RuntimeError("No sandbox")
    return dict(zip(SCORE_FIELDS, _attempt_key(draft, result), strict=True))


def _impact_json(impact: PreviewImpact) -> dict[str, Any]:
    return {
        "new_interdits": [_warning_json(item) for item in impact.new_interdits],
        "broken_wishes": [_warning_json(item) for item in impact.broken_wishes],
        "contract": [
            {
                "employee_id": row.employee_id,
                "week_start": row.week_start,
                "current_hours": row.current_hours,
                "trial_hours": row.trial_hours,
                "contracted": row.contracted,
                "kind": row.kind,
            }
            for row in impact.contract
        ],
        "coverage_added": [_warning_json(item) for item in impact.coverage_added],
        "coverage_removed": [_warning_json(item) for item in impact.coverage_removed],
        "role_fit": [
            {
                "current_gap": row.current_gap,
                "trial_gap": row.trial_gap,
                "kind": row.kind,
            }
            for row in impact.role_fit
        ],
    }


def _slot_json(slot: FillSlot) -> dict[str, Any]:
    return {
        "employee_id": slot.employee_id,
        "day_index": slot.day_index,
        "weekday": slot.weekday,
        "service_id": slot.service_id,
        "team": slot.team.value,
    }


def _recap_json(gesture: str, body: dict[str, Any], chosen: PreviewProposal, index: int) -> dict[str, Any]:
    occupied = gesture in {"retune", "replace", "swap"}
    shift = parse_shift(body.get("shift") or {}) if occupied else None
    slot_raw = body.get("slot")
    slot = parse_slot(slot_raw) if gesture == "fill" and isinstance(slot_raw, dict) else None
    return {
        "index": index,
        "gesture": gesture,
        "shift": None if shift is None else _shift_json(shift),
        "slot": None if slot is None else _slot_json(slot),
        "employee_id": chosen.employee_id,
        "start_minutes": chosen.start_minutes,
        "end_minutes": chosen.end_minutes,
        "partner": None if chosen.partner is None else _shift_json(chosen.partner),
        "impact": _impact_json(chosen.impact),
    }


def _proposal_json(item: PreviewProposal) -> dict[str, Any]:
    return {
        "rank": item.rank,
        "gesture": item.gesture,
        "start_minutes": item.start_minutes,
        "end_minutes": item.end_minutes,
        "employee_id": item.employee_id,
        "partner": None if item.partner is None else _shift_json(item.partner),
        "impact": _impact_json(item.impact),
        "current_score": dict(zip(SCORE_FIELDS, item.current_score, strict=True)),
        "trial_score": dict(zip(SCORE_FIELDS, item.trial_score, strict=True)),
    }


def _warning_from_json(raw: dict[str, Any]) -> Warning:
    return Warning(
        WarningSeverity(raw["severity"]),
        raw["code"],
        raw["message"],
        raw.get("employee_id"),
        raw.get("day_index"),
    )


def _persist() -> None:
    from doux_planning.api.db import SandboxSession, database_url, session_scope

    if not database_url():
        return
    sandbox = _sandbox_or_none()
    if sandbox is None:
        return
    result = sandbox.last_result
    document = {
        "recaps": list(_recaps),
        "assignments": [_shift_json(item) for item in sandbox.draft.assignments],
        "warnings": [_warning_json(item) for item in (result.warnings if result else ())],
        "history": [
            {
                "assignments": [_shift_json(item) for item in snap.assignments],
                "warnings": [
                    _warning_json(item)
                    for item in (snap.last_result.warnings if snap.last_result is not None else ())
                ],
            }
            for snap in sandbox.history
        ],
    }
    with session_scope() as session:
        row = session.get(SandboxSession, RESTAURANT_ID)
        if row is None:
            session.add(SandboxSession(restaurant_id=RESTAURANT_ID, document=document))
        else:
            row.document = document
            flag_modified(row, "document")


def _delete_persisted() -> None:
    from doux_planning.api.db import SandboxSession, database_url, session_scope

    if not database_url():
        return
    try:
        with session_scope() as session:
            row = session.get(SandboxSession, RESTAURANT_ID)
            if row is not None:
                session.delete(row)
    except Exception:
        return


def _restore() -> None:
    from doux_planning.api.db import SandboxSession, database_url, session_scope

    global _recaps
    if not database_url() or _store is None:
        return
    try:
        with session_scope() as session:
            row = session.get(SandboxSession, RESTAURANT_ID)
            document = None if row is None else dict(row.document)
    except Exception:
        return
    if not document:
        return
    hydrate_delivered_cycle(_store, RESTAURANT_ID)
    sandbox = _store.get(RESTAURANT_ID).sandbox
    assert sandbox is not None
    assignments = tuple(parse_shift(item) for item in document["assignments"])
    warnings = tuple(_warning_from_json(item) for item in document.get("warnings") or ())
    sandbox.draft = sandbox.draft.with_assignments(assignments)
    sandbox.last_result = EngineResult(assignments=assignments, warnings=warnings)
    sandbox.history = [
        SandboxSnapshot(
            assignments=tuple(parse_shift(item) for item in snap["assignments"]),
            last_result=EngineResult(
                assignments=tuple(parse_shift(item) for item in snap["assignments"]),
                warnings=tuple(_warning_from_json(item) for item in snap.get("warnings") or ()),
            ),
        )
        for snap in document.get("history") or []
    ]
    _recaps = list(document.get("recaps") or [])

File: web/src/sandbox.ts
import {
  isRecord,
  parseAssignment,
  parseEmployee,
  parseWarning,
  PayloadError,
  requireArray,
  requireNumber,
  requireRecord,
  requireString,
} from "./api";
import type {
  Assignment,
  ContractImpact,
  ContractKind,
  FillSlot,
  Gesture,
  HistoryCran,
  HistoryEntry,
  Impact,
  PreviewProposal,
  RoleFit,
  RoleFitKind,
  SandboxState,
  Score,
  ShiftIdentity,
} from "./types";
import { SCORE_FIELDS, toShiftIdentity } from "./types";

export class ApiHttpError extends Error {
  constructor(
    public status: number,
    public detail: string,
  ) {
    super(detail);
    this.name = "ApiHttpError";
  }
}

async function parseDetail(response: Response): Promise<string> {
  try {
    const body: unknown = await response.json();
    if (isRecord(body) && typeof body.detail === "string") {
      return body.detail;
    }
  } catch {
    /* ignore */
  }
  return `HTTP ${response.status}`;
}

async function sendJson(url: string, init?: RequestInit): Promise<unknown> {
  const response = await fetch(url, init);
  if (!response.ok) {
    throw new ApiHttpError(response.status, await parseDetail(response));
  }
  return response.json();
}

function parseGesture(value: unknown, path: string): Gesture {
  if (value === "retune" || value === "replace" || value === "swap" || value === "fill") {
    return value;
  }
  throw new PayloadError(`geste inattendu : ${path}`);
}

function parseNullableNumber(value: unknown, path: string): number | null {
  if (value === null) {
    return null;
  }
  if (typeof value !== "number" || Number.isNaN(value)) {
    throw new PayloadError(`clé invalide : ${path}`);
  }
  return value;
}

function parseNullableString(value: unknown, path: string): string | null {
  if (value === null) {
    return null;
  }
  if (typeof value !== "string") {
    throw new PayloadError(`clé invalide : ${path}`);
  }
  return value;
}

function parseScore(value: unknown, path: string): Score {
  if (!isRecord(value)) {
    throw new PayloadError(`objet attendu : ${path}`);
  }
  const score = {} as Score;
  for (const field of SCORE_FIELDS) {
    score[field] = requireNumber(value, field, path);
  }
  return score;
}

function parseContractKind(value: unknown, path: string): ContractKind {
  if (value === "closer" || value === "farther" || value === "excess") {
    return value;
  }
  throw new PayloadError(`kind inattendu : ${path}`);
}

function parseRoleFitKind(value: unknown, path: string): RoleFitKind {
  if (value === "better" || value === "worse") {
    return value;
  }
  throw new PayloadError(`kind inattendu : ${path}`);
}

function parseWeekStart(value: unknown, path: string): 0 | 7 {
  if (value === 0 || value === 7) {
    return value;
  }
  throw new PayloadError(`week_start inattendu : ${path}`);
}

function parseContractImpact(value: unknown, path: string): ContractImpact {
  if (!isRecord(value)) {
    throw new PayloadError(`objet attendu : ${path}`);
  }
  return {
    employee_id: requireString(value, "employee_id", path),
    week_start: parseWeekStart(value.week_start, `${path}.week_start`),
    current_hours: requireNumber(value, "current_hours", path),
    trial_hours: requireNumber(value, "trial_hours", path),
    contracted: requireNumber(value, "contracted", path),
    kind: parseContractKind(value.kind, `${path}.kind`),
  };
}

function parseRoleFit(value: unknown, path: string): RoleFit {
  if (!isRecord(value)) {
    throw new PayloadError(`objet attendu : ${path}`);
  }
  return {
    current_gap: requireNumber(value, "current_gap", path),
    trial_gap: requireNumber(value, "trial_gap", path),
    kind: parseRoleFitKind(value.kind, `${path}.kind`),
  };
}

function parseImpact(value: unknown, path: string): Impact {
  if (!isRecord(value)) {
    throw new PayloadError(`objet attendu : ${path}`);
  }
  return {
    new_interdits: requireArray(value, "new_interdits", path).map((item, i) =>
      parseWarning(item, `${path}.new_interdits[${i}]`),
    ),
    broken_wishes: requireArray(value, "broken_wishes", path).map((item, i) =>
      parseWarning(item, `${path}.broken_wishes[${i}]`),
    ),
    contract: requireArray(value, "contract", path).map((item, i) =>
      parseContractImpact(item, `${path}.contract[${i}]`),
    ),
    coverage_added: requireArray(value, "coverage_added", path).map((item, i) =>
      parseWarning(item, `${path}.coverage_added[${i}]`),
    ),
    coverage_removed: requireArray(value, "coverage_removed", path).map((item, i) =>
      parseWarning(item, `${path}.coverage_removed[${i}]`),
    ),
    role_fit: requireArray(value, "role_fit", path).map((item, i) => parseRoleFit(item, `${path}.role_fit[${i}]`)),
  };
}

function parseProposal(value: unknown, path: string): PreviewProposal {
  if (!isRecord(value)) {
    throw new PayloadError(`objet attendu : ${path}`);
  }
  if (!("start_minutes" in value) || !("end_minutes" in value) || !("employee_id" in value) || !("partner" in value)) {
    throw new PayloadError(`clé absente : ${path}`);
  }
  const partner = value.partner;
  if (partner !== null && !isRecord(partner)) {
    throw new PayloadError(`clé invalide : ${path}.partner`);
  }
  return {
    rank: requireNumber(value, "rank", path),
    gesture: parseGesture(value.gesture, `${path}.gesture`),
    start_minutes: parseNullableNumber(value.start_minutes, `${path}.start_minutes`),
    end_minutes: parseNullableNumber(value.end_minutes, `${path}.end_minutes`),
    employee_id: parseNullableString(value.employee_id, `${path}.employee_id`),
    partner: partner === null ? null : parseAssignment(partner, `${path}.partner`),
    impact: parseImpact(value.impact, `${path}.impact`),
    current_score: parseScore(value.current_score, `${path}.current_score`),
    trial_score: parseScore(value.trial_score, `${path}.trial_score`),
  };
}

function parseServiceId(value: unknown, path: string): "midday" | "evening" {
  if (value === "midday" || value === "evening") {
    return value;
  }
  throw new PayloadError(`service_id inattendu : ${path}`);
}

function parseFillSlot(value: unknown, path: string): FillSlot {
  if (!isRecord(value)) {
    throw new PayloadError(`objet attendu : ${path}`);
  }
  return {
    employee_id: requireString(value, "employee_id", path),
    day_index: requireNumber(value, "day_index", path),
    weekday: requireString(value, "weekday", path),
    service_id: parseServiceId(value.service_id, `${path}.service_id`),
    team: requireString(value, "team", path),
  };
}

function parseNullableAssignment(value: unknown, path: string): Assignment | null {
  if (value === null) {
    return null;
  }
  return parseAssignment(value, path);
}

function parseNullableFillSlot(value: unknown, path: string): FillSlot | null {
  if (value === null) {
    return null;
  }
  return parseFillSlot(value, path);
}

function parseHistoryCran(value: unknown, path: string): HistoryCran {
  if (!isRecord(value)) {
    throw new PayloadError(`objet attendu : ${path}`);
  }
  if (
    !("shift" in value) ||
    !("slot" in value) ||
    !("employee_id" in value) ||
    !("start_minutes" in value) ||
    !("end_minutes" in value) ||
    !("partner" in value)
  ) {
    throw new PayloadError(`clé absente : ${path}`);
  }
  return {
    index: requireNumber(value, "index", path),
    gesture: parseGesture(value.gesture, `${path}.gesture`),
    shift: parseNullableAssignment(value.shift, `${path}.shift`),
    slot: parseNullableFillSlot(value.slot, `${path}.slot`),
    employee_id: parseNullableString(value.employee_id, `${path}.employee_id`),
    start_minutes: parseNullableNumber(value.start_minutes, `${path}.start_minutes`),
    end_minutes: parseNullableNumber(value.end_minutes, `${path}.end_minutes`),
    partner: parseNullableAssignment(value.partner, `${path}.partner`),
    impact: parseImpact(value.impact, `${path}.impact`),
  };
}

const SILENT_SCORE: Score = {
  empty: 0,
  interdit: 0,
  hours_miss: 0,
  souhait: 0,
  below_role: 0,
  overqualification: 0,
};

export function historyEntryFromCran(cran: HistoryCran): HistoryEntry {
  return {
    index: cran.index,
    gesture: cran.gesture,
    shift: cran.shift ? toShiftIdentity(cran.shift) : null,
    slot: cran.slot,
    proposal: {
      rank: 1,
      gesture: cran.gesture,
      start_minutes: cran.start_minutes,
      end_minutes: cran.end_minutes,
      employee_id: cran.employee_id,
      partner: cran.partner,
      impact: cran.impact,
      current_score: SILENT_SCORE,
      trial_score: SILENT_SCORE,
    },
  };
}

export function parseSandboxState(value: unknown): SandboxState {
  if (!isRecord(value)) {
    throw new PayloadError("réponse sandbox invalide");
  }
  const sandbox = requireRecord(value, "sandbox", "root");
  const restaurant = requireRecord(value, "restaurant", "root");
  const planning = requireRecord(value, "planning", "root");
  return {
    sandbox: {
      target: requireString(sandbox, "target", "sandbox"),
      history_length: requireNumber(sandbox, "history_length", "sandbox"),
    },
    restaurant: {
      id: requireString(restaurant, "id", "restaurant"),
      name: requireString(restaurant, "name", "restaurant"),
      employees: requireArray(restaurant, "employees", "restaurant").map((item, i) =>
        parseEmployee(item, `restaurant.employees[${i}]`),
      ),
    },
    planning: {
      assignments: requireArray(planning, "assignments", "planning").map((item, i) =>
        parseAssignment(item, `planning.assignments[${i}]`),
      ),
      warnings: requireArray(planning, "warnings", "planning").map((item, i) =>
        parseWarning(item, `planning.warnings[${i}]`),
      ),
    },
    score: parseScore(value.score, "score"),
    history: requireArray(value, "history", "root").map((item, i) => parseHistoryCran(item, `history[${i}]`)),
  };
}

export function parsePreviewBody(value: unknown): { proposals: PreviewProposal[] } {
  if (!isRecord(value)) {
    throw new PayloadError("réponse preview invalide");
  }
  return {
    proposals: requireArray(value, "proposals", "root").map((item, i) => parseProposal(item, `proposals[${i}]`)),
  };
}

export async function enterSandbox(): Promise<SandboxState> {
  return parseSandboxState(await sendJson("/v1/sandbox/enter", { method: "POST" }));
}

export async function getSandbox(): Promise<SandboxState> {
  return parseSandboxState(await sendJson("/v1/sandbox"));
}

export async function previewSandbox(
  gesture: Gesture,
  shift: ShiftIdentity,
  hours?: { start_minutes: number; end_minutes: number },
): Promise<PreviewProposal[]> {
  const payload: Record<string, unknown> = { gesture, shift };
  if (gesture === "retune") {
    if (!hours) {
      throw new PayloadError("start_minutes et end_minutes requis pour retune");
    }
    payload.start_minutes = hours.start_minutes;
    payload.end_minutes = hours.end_minutes;
  }
  const body = parsePreviewBody(
    await sendJson("/v1/sandbox/preview", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }),
  );
  return body.proposals;
}

export async function previewFill(
  slot: FillSlot,
  hours: { start_minutes: number | null; end_minutes: number | null },
): Promise<PreviewProposal[]> {
  const body = parsePreviewBody(
    await sendJson("/v1/sandbox/preview", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        gesture: "fill",
        slot,
        start_minutes: hours.start_minutes,
        end_minutes: hours.end_minutes,
      }),
    }),
  );
  return body.proposals;
}

export async function commitSandbox(body: Record<string, unknown>): Promise<SandboxState> {
  return parseSandboxState(
    await sendJson("/v1/sandbox/commit", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
  );
}

export async function undoSandbox(): Promise<SandboxState> {
  return parseSandboxState(await sendJson("/v1/sandbox/undo", { method: "POST" }));
}

export async function discardSandbox(): Promise<SandboxState> {
  return parseSandboxState(await sendJson("/v1/sandbox/discard", { method: "POST" }));
}

export function commitFillBody(
  slot: FillSlot,
  proposal: PreviewProposal,
): Record<string, unknown> {
  return {
    gesture: "fill",
    slot,
    employee_id: proposal.employee_id,
    start_minutes: proposal.start_minutes,
    end_minutes: proposal.end_minutes,
  };
}

export function commitBody(gesture: Gesture, shift: ShiftIdentity, proposal: PreviewProposal): Record<string, unknown> {
  const payload: Record<string, unknown> = { gesture, shift };
  if (gesture === "retune") {
    payload.start_minutes = proposal.start_minutes;
    payload.end_minutes = proposal.end_minutes;
  } else if (gesture === "replace") {
    payload.employee_id = proposal.employee_id;
  } else if (gesture === "swap") {
    payload.partner = proposal.partner
      ? {
          employee_id: proposal.partner.employee_id,
          day_index: proposal.partner.day_index,
          weekday: proposal.partner.weekday,
          service_id: proposal.partner.service_id,
          team: proposal.partner.team,
          start_minutes: proposal.partner.start_minutes,
          end_minutes: proposal.partner.end_minutes,
          post_level: proposal.partner.post_level,
        }
      : null;
  }
  return payload;
}
